iavl/iavl_proof2.go

```diff
@@ -167,52 +171,255 @@ func (proof *KeyNotExistsProof) Verify(key []byte, root []byte) error {
 	return nil
 }
 
-type KeyRangeExistsProof struct {
-	RootHash   data.Bytes
-	PathToKeys []*PathToKey
+// KeyRangeProof is proof that a range of keys does or does not exist.
+type KeyRangeProof struct {
+	RootHash   data.Bytes   `json:"root_hash"`
+	PathToKeys []*PathToKey `json:"paths"`
+
+	LeftPath *PathToKey        `json:"left_path"`
+	LeftNode IAVLProofLeafNode `json:"left_node"`
+
+	RightPath *PathToKey        `json:"right_path"`
+	RightNode IAVLProofLeafNode `json:"right_node"`
+}
+
+func (proof *KeyRangeProof) String() string {
+	paths := []string{}
+	for _, p := range proof.PathToKeys {
+		paths = append(paths, "\t"+p.String())
+	}
+	inner := fmt.Sprintf("PathToKeys: \n\t%s", strings.Join(paths, ",\n"))
+	inner += fmt.Sprintf("LeftNode: %#v (%x)\n", proof.LeftNode, proof.LeftNode.Hash())
+	inner += fmt.Sprintf("RightNode: %#v (%x)", proof.RightNode, proof.RightNode.Hash())
+	return "&KeyRangeProof{\n\t" + inner + "\n}"
+}
+
+// Returns a list of all paths, in order.
+func (proof *KeyRangeProof) paths() []*PathToKey {
+	paths := proof.PathToKeys[:]
+	if proof.LeftPath != nil {
+		paths = append([]*PathToKey{proof.LeftPath}, paths...)
+	}
+	if proof.RightPath != nil {
+		paths = append(paths, proof.RightPath)
+	}
+	return paths
+}
+
+// Checks that all paths are adjacent to one another, with no gaps.
+func (proof *KeyRangeProof) verifyPathAdjacency() error {
+	paths := proof.paths()
+	for i, path := range paths {
+		if i >= len(paths)-1 {
```

Comment:
I think it would be better to use a `for` loop here: `for i := 0; i < len(paths)-1; i++` m?